crypto/thresholdsign.go

```diff
@@ -13,40 +13,150 @@ import (
+type ThresholdSignatureInspector interface {
+	// VerifyShare verifies the input signature against the stored message and stored
+	// key at the input index. This function does not update the internal state.
+	// The function is thread-safe.
+	// Returns:
+	//  - (true, nil) if the signature is valid
+	//  - (false, nil) if `orig` is a valid index but the signature share is invalid
+	//  - (false, InvalidInputsError) if `orig` is an invalid index value
+	//  - (false, error) for all other unexpected errors
+	VerifyShare(orig int, share Signature) (bool, error)
+
+	// VerifyThresholdSignature verifies the input signature against the stored
+	// message and stored group public key. It does not update the internal state.
+	// The function is thread-safe.
+	// Returns:
+	//  - (true, nil) if the signature is valid
+	//  - (false, nil) if the signature is invalid
+	//  - (false, error) for all other unexpected errors
+	VerifyThresholdSignature(thresholdSignature Signature) (bool, error)
+
+	// EnoughShares indicates whether enough shares have been accumulated in order to reconstruct
+	// a group signature. This function is thread safe and locks the internal state.
+	// Returns:
+	//  - true if and only if at least (threshold+1) shares were added
+	EnoughShares() bool
+
+	// TrustedAdd adds a signature share to the internal pool of shares
+	// without verifying the signature against the message and the participant's
+	// public key. This function is thread safe and locks the internal state.
+	//
+	// The share is only added if the signer index is valid and has not been
+	// added yet. Moreover, the share is added only if not enough shares were collected.
+	// The function returns:
+	//  - (true, nil) if enough signature shares were already collected and no error occured
+	//  - (false, nil) if not enough shares were collected and no error occured
+	//  - (false, InvalidInputsError) if index is invalid
+	//  - (false, duplicatedSignerError) if a signature for the index was previously added
+	TrustedAdd(orig int, share Signature) (bool, error)
+
+	// VerifyAndAdd verifies a signature share (same as `VerifyShare`),
+	// and may or may not add the share to the local pool of shares.
+	// This function is thread safe and locks the internal state.
+	//
+	// The share is only added if the signature is valid, the signer index is valid and has not been
+	// added yet. Moreover, the share is added only if not enough shares were collected.
+	// The function returns 3 outputs:
+	//  - First boolean output is true if the share is valid and no error is returned, and false otherwise.
+	//  - Second boolean output is true if enough shares were collected and no error is returned, and false otherwise.
+	//  - error is InvalidInputsError if input index is invalid, duplicatedSignerError if signer was added,
+	//	   and a random error if an exception occurred.
+	//    (an invalid signature is not considered an invalid input, look at `VerifyShare` for details)
```

Comment:
I think we can increase readability here by converting the errors to a bullet point list:
```suggestion
	// Boolean returns:
	//  - First boolean output is true if the share is valid and no error is returned, and false otherwise.
	//  - Second boolean output is true if enough shares were collected and no error is returned, and false otherwise.
	// Error returns:
	//  - InvalidInputsError if input index is invalid
    //    (invalid signature will not cause an error here, as the signature is not checked yet)
	//  - duplicatedSignerError if signature from orig was previously added,
	// All other errors are unexpected during normal operation and a potential symptom of internal state corruption. 
```